lib/core/src/Cardano/Wallet/Primitive/Types/MinimumUTxO/Gen.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+-- Defines generators and shrinkers for the 'MinimumUTxO' data type.
+--
+module Cardano.Wallet.Primitive.Types.MinimumUTxO.Gen
+    ( genMinimumUTxO
+    , genMinimumUTxOForShelleyBasedEra
+    , shrinkMinimumUTxO
+    , shrinkMinimumUTxOForShelleyBasedEra
+    )
+    where
+
+import Prelude
+
+import Cardano.Api
+    ( ShelleyBasedEra (..) )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.MinimumUTxO
+    ( MinimumUTxO (..), MinimumUTxOForShelleyBasedEra (..) )
+import Data.Bits
+    ( Bits )
+import Data.Default
+    ( Default (..) )
+import Data.IntCast
+    ( intCast, intCastMaybe )
+import Data.Maybe
+    ( fromMaybe )
+import Numeric.Natural
+    ( Natural )
+import Test.QuickCheck
+    ( Gen, choose, frequency, oneof )
+
+import qualified Cardano.Ledger.Alonzo.PParams as Alonzo
+import qualified Cardano.Ledger.Babbage.PParams as Babbage
+import qualified Cardano.Ledger.Coin as Ledger
+import qualified Cardano.Ledger.Shelley.PParams as Shelley
+
+--------------------------------------------------------------------------------
+-- Generating 'MinimumUTxO' values
+--------------------------------------------------------------------------------
+
+genMinimumUTxO :: Gen MinimumUTxO
+genMinimumUTxO = frequency
+    [ (1, genMinimumUTxONone)
+    , (1, genMinimumUTxOConstant)
+    , (8, MinimumUTxOForShelleyBasedEraOf <$> genMinimumUTxOForShelleyBasedEra)
+    ]
+  where
+    genMinimumUTxONone :: Gen MinimumUTxO
+    genMinimumUTxONone = pure MinimumUTxONone
+
+    genMinimumUTxOConstant :: Gen MinimumUTxO
+    genMinimumUTxOConstant = MinimumUTxOConstant . Coin
+        <$> genInterestingCoinValue
```

Comment:
Hi @Anviking

I had a look for `genCoin` within `Cardano.Api.Gen`, but could only find `genLovelace`. I'm guessing you mean this function?

According to the current definition of `genLovelace`, it has a chance of generating values close to `txOutMaxLovelace`:

```hs
genLovelace :: Gen Lovelace
genLovelace = frequency
    [ (10, Lovelace . intCast . getNonNegative @Int <$> arbitrary)
    , (50, choose (1_000_000, 1_000_000_000))
    , (10, choose (txOutMinLovelace, txOutMaxLovelace))
    , (30, genEncodingBoundaryLovelace)
    ]
```

But for this test case, the values we generate will ultimately be multiplied by the number of bytes (or words) in the serialized representation of a `TxOut` in order to calculate the minimum ada quantity. So if we use this generator, we could end up computing minimum ada quantities that are greater than `txOutMaxLovelace`.

Reply:
Hi @Anviking 

After reading your comment, I spent a bit of time thinking about the existing `genInterestingLedgerCoin` generator.

One problem with this generator is that we reuse it for all eras within `genMinimumUTxO`. But the `Coin` value has a different meaning in each era, and realistic values have different magnitudes in different eras.

For example, on average, we would expect that these values to have different magnitudes from one another:
   - the `coinsPerUTxOWord` parameter (Alonzo era)
   - the `coinsPerUTxOByte` parameter (Babbage era)

With this in mind, I think it would be better to generate values that are appropriate to the era. And to make them more realistic, we can generate values that are of a similar magnitude to real-life `mainnet` protocol parameters.

I've made this change in 114cf90e0b8be67b730b9c296f31f2a49a5ec4d8.